File: android/java/org/chromium/chrome/browser/crypto_wallet/KeyringServiceFactory.java
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.brave_wallet.mojom.KeyringService;
import org.chromium.chrome.browser.crypto_wallet.util.Utils;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.MojoException;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
public class KeyringServiceFactory {
    private static final Object lock = new Object();
    private static KeyringServiceFactory instance;

    public static KeyringServiceFactory getInstance() {
        synchronized (lock) {
            if (instance == null) {
                instance = new KeyringServiceFactory();
            }
        }
        return instance;
    }

    public KeyringService getKeyringService(ConnectionErrorHandler connectionErrorHandler) {
        Profile profile = Utils.getProfile(false); // Always use regular profile
        int nativeHandle = KeyringServiceFactoryJni.get().getInterfaceToKeyringService(profile);
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        KeyringService keyringService = KeyringService.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) keyringService).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return keyringService;
    }

    private MessagePipeHandle wrapNativeHandle(int nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        int getInterfaceToKeyringService(Profile profile);
    }
}

File: browser/brave_wallet/keyring_service_factory.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "brave/browser/brave_wallet/keyring_service_factory.h"

#include <utility>

#include "brave/browser/brave_wallet/brave_wallet_context_utils.h"
#include "brave/browser/brave_wallet/json_rpc_service_factory.h"
#include "brave/components/brave_wallet/browser/keyring_service.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"
#include "components/user_prefs/user_prefs.h"

namespace brave_wallet {

// static
KeyringServiceFactory* KeyringServiceFactory::GetInstance() {
  return base::Singleton<KeyringServiceFactory>::get();
}

// static
mojo::PendingRemote<mojom::KeyringService> KeyringServiceFactory::GetForContext(
    content::BrowserContext* context) {
  if (!IsAllowedForContext(context))
    return mojo::PendingRemote<mojom::KeyringService>();

  return static_cast<KeyringService*>(
             GetInstance()->GetServiceForBrowserContext(context, true))
      ->MakeRemote();
}

// static
KeyringService* KeyringServiceFactory::GetServiceForContext(
    content::BrowserContext* context) {
  if (!IsAllowedForContext(context)) {
    return nullptr;
  }
  return static_cast<KeyringService*>(
      GetInstance()->GetServiceForBrowserContext(context, true));
}

// static
void KeyringServiceFactory::BindForContext(
    content::BrowserContext* context,
    mojo::PendingReceiver<mojom::KeyringService> receiver) {
  auto* keyring_service = KeyringServiceFactory::GetServiceForContext(context);
  if (keyring_service) {
    keyring_service->Bind(std::move(receiver));
  }
}

KeyringServiceFactory::KeyringServiceFactory()
    : BrowserContextKeyedServiceFactory(
          "KeyringService",
          BrowserContextDependencyManager::GetInstance()) {
  DependsOn(JsonRpcServiceFactory::GetInstance());
}

KeyringServiceFactory::~KeyringServiceFactory() = default;

KeyedService* KeyringServiceFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  return new KeyringService(
      JsonRpcServiceFactory::GetServiceForContext(context),
      user_prefs::UserPrefs::Get(context), g_browser_process->local_state());
}

content::BrowserContext* KeyringServiceFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

}  // namespace brave_wallet
